composer/trainer/checkpoint.py

```diff
@@ -415,23 +443,39 @@ def save_checkpoint(self, state: State, seed: int, device: Device) -> None:
             if is_module_deepspeed(state.model):
                 model = cast("deepspeed.DeepSpeedEngine", state.model)
                 model.save_checkpoint(tmpdir, _DEEPSPEED_TAG)
+                # ensure that deepspeed checkpoints are saved in an archive
+                self.file_extension, self.write_mode = _ensure_archive(file_extension=self.file_extension,
+                                                                       write_mode=self.write_mode)
 
+            composer_states_filepath = os.path.join(tmpdir, _COMPOSER_STATES_FILENAME)
             if dist.get_global_rank() == 0:
                 # only rank 0 saves checkpoints
 
                 # we add the state only on rank 0 since other processes don't have loggers to serialize
                 # it should be the same across all ranks. per-rank state not stored
                 state_dict['state'] = state.state_dict()
 
-                composer_states_filepath = os.path.join(tmpdir, _COMPOSER_STATES_FILENAME)
                 with open(composer_states_filepath, 'xb') as f:
                     torch.save(state_dict, f)
 
-            checkpoint_archive_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
-            with tarfile.open(checkpoint_archive_filepath, self.write_mode) as tarball:
-                tarball.add(tmpdir, arcname="")  # add files flat to the tarball
-
-            log.info(f'Trainer checkpoint saved to {checkpoint_archive_filepath}')
+            checkpoint_filepath = os.path.join(self.checkpoint_folder, f'{tag}{self.file_extension}')
+            if _is_pt_file(checkpoint_filepath):
+                # only rank 0 will have save states to a pt file
+                if dist.get_global_rank() == 0:
+                    # move the file out of tmpdir to the user-specified location
+                    shutil.move(composer_states_filepath, checkpoint_filepath)
+            else:
+                # using compression, deepspeed, or both
+                with tarfile.open(checkpoint_filepath, self.write_mode) as tarball:
+                    # add files flat to the tarball with the specified compression
+                    tarball.add(tmpdir, arcname="")
+
+            timestamp = state.timer.get_timestamp()
+            paths = dist.all_gather_object(checkpoint_filepath) if is_module_deepspeed(
```

Comment:
Instead of using a distributed reduction here, I'm wondering if it might be better if I backtrack on my original comment, and instead have each checkpoint saver keep track only of the files it wrote. I'm now realizing that this could be misleading to the user in a multi-node environment, since this function would store paths to files on other nodes. Do you think it would be better if we offloaded this responsibility to the user -- i.e. if you want all checkpoint file paths, perform your own distributed reduction?

Otherwise, I think we should make this extremely clear in the docstring -- perhaps using `.. note::` notation -- and explain that the list index corresponds to the rank on which that the filepath is valid?

Reply:
We should make it as easy as possible for the user to get the checkpoint paths so ideally they don't have to do the reduction themselves. I'll make it extremely clear in the docstring.